Correct inconsistencies and errors in Nav Editor docs

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Following as discussion with @talldan I discovered that there are some inconsistencies in the Nav Editor docs. This PR corrects these. See individual commits for specifics.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Explain that the entire navigatoin is rendered as blocks not just the non link blocks

talldan commented on packages/edit-navigation/README.md:
```diff
@@ -31,7 +31,9 @@ The editing experience is provided as a block editor wrapper around the core fun
 The Navigation Editor has two "modes" for _persistence_ ("saving" navigations) and _rendering_:
 
 1. **Default** - navigations are saved to the _existing_ (post type powered) Menus system and rendered using standard Walker classes.
-2. **Block-based** (opt _in_) - navigations continue to be _saved_ using the existing post type system, but non-link blocks are saved (see technical implementation) and _rendered_ as blocks to provide access to the full power of the Navigation block (with some tradeoffs in terms of backwards compatibility).
+2. **Block-based** (opt _in_) - navigations continue to be _saved_ using the existing post type system, but non-link blocks are
+    - saved (see technical implementation) as blocks.
+    - the navigation is rendered using the Navigation block (as opposed to Walker) to provide access to the full power of the Navigation block (with some tradeoffs in terms of backwards compatibility).
```
I think it might be worth referring more to the 'links' rather than the navigation as a whole, since that's really where the magic happens.

1. **Default** - navigation links are converted to and saved using the _existing_ menu item post type system and rendered using standard Walker classes.
2. **Block-based** (opt _in_) - navigation links continue to be _saved_ using the existing menu item post type system, but non-link blocks are
    - saved as block HTML in the menu item post type (see the technical implementation section).
    - the entire navigation is rendered using the Navigation block as opposed to Walker (with some tradeoffs in terms of backwards compatibility). Menu items that are blocks are rendered as blocks. Menu items that are links are rendered as Navigation Link blocks.

getdave replied:
I think we should disambiguate `links` as a term. So

* Navigation item - the individual items. Can be of any type.
* Navigation Link - refers to the `core/navigation-link` block.

Do we want to drop the term `Menu` entirely? I feel like Matias suggested leaning towards navigation(s)...etc.

talldan replied:
Yeah, that sounds fair. I only really use menu, when referring to the something related to the Menu screen as it's called currently in WordPress.

getdave committed:
Update to increase clarity of sentance

getdave committed:
Add notes about backwards compatibility

getdave committed:
Improve backwards compat docs

talldan reviewed (approved):
Thanks for noticing this and updating it 🙇

getdave committed:
Correct things to document that block-based links are still rendered on the. front end

getdave committed:
Updated glossary with block-based link info

getdave commented:
> Thanks for noticing this and updating it 🙇

@talldan Just noticed your 👍 here. I made some additional changes. Are you still happy?

talldan commented on packages/edit-navigation/README.md:
```diff
@@ -61,9 +61,14 @@ This unlocks significant additional capabilities in the Navigation Editor. For e
 
-_Non_-link navigation items however, are [persisted as `nav_menu_items` with a special `type` of `block`](https://github.com/WordPress/gutenberg/blob/7fcd57c9a62c232899e287f6d96416477d810d5e/packages/edit-navigation/src/store/utils.js#L159-L166). These items have an [_additional_ `content` field which is used to store the serialized block markup](https://github.com/WordPress/gutenberg/blob/7fcd57 ...
+1. Navigation links - these utilise the `core/navigation-link` block and are _not_ persisted as blocks.
+2. Block-based links - these are any item that is _not_ a `core/navigation-link` block. They are persisted as blocks (via a wrapper - see below).
```
Using the terminology 'links' here might be confusing as they're often not links (spacer, search etc.)

getdave replied:
I'll look into this.

getdave replied:
Fixed in https://github.com/WordPress/gutenberg/pull/34682/commits/d6bdb6ef25306e689c6313123964b052b2a25e5b

draganescu commented on packages/edit-navigation/README.md:
```diff
@@ -59,13 +61,26 @@ This unlocks significant additional capabilities in the Navigation Editor. For e
 #### Technical Implementation details
 
-By default, `core/navigation-link` items are serialized and persisted as `nav_menu_item` posts. No serialized block HTML is stored for these standard link blocks.
+There are two types of item that can be added to a navigation:
```
I wouldn't say there are only two things that can be added to a navigation because it is not true. I would say there are two kinds of navigations: classic, made of links, and block based made of blocks.

Because the editor is a block editor, we represent a classic link with a `core/navigation-link` A `core/navigation-link` block also represents a link when editing a block based menu.

This logic then applies below to the explanation: instead of the dichotomy "block based link" and "navigation link", use classic and block based contexts to explain what happens to the blocks in the editor, _normally_ when editing a classic menu and  _extra_ when editing a block based menu.

talldan replied:
> I would say there are two kinds of navigations: classic, made of links, and block based made of blocks.

Though even in a block based menu navigation link and navigation submenu blocks are saved in a different way to other blocks.

getdave replied:
Exactly. The idea is to make it clear that (currently) even if you used other types of blocks (ie: not `core/navigatoin-link`) the block data is stored using the existing menu system.

That might change but right now it is how it is.

getdave replied:
I really think the purpose of this section of the docs is to explain the difference between the types of _item_. Earlier in the docs I go into a lot of detail about the two "modes"

1. Classic
2. Block based.

I don't think we ought to repeat this again here.

draganescu replied:
The more I thought about explaining the more I think we should skip over documenting something which is an experiment that we'll definitely deprecate (I mean the saving of blocks like HTML menu items)

getdave replied:
Right so basically we need to remove the entire `Technical Implementation details` section.

getdave replied:
Ok done. Noting that this PR was just intending to correct inconsistent information and errors so I'd rather not remove or rewrite any other large sections of docs within this PR if that's ok?

draganescu commented on packages/edit-navigation/README.md:
```diff
@@ -59,13 +61,26 @@ This unlocks significant additional capabilities in the Navigation Editor. For e
+
+By default, Navigation link items are serialized and persisted as `nav_menu_item` posts. No serialized block HTML is stored for these standard link blocks.
+
+Block-based link items however, are [persisted as `nav_menu_items` with a special `type` of `block`](https://github.com/WordPress/gutenberg/blob/7fcd57c9a62c232899e287f6d96416477d810d5e/packages/edit-navigation/src/store/utils.js#L159-L166). These items have an [_additional_ `content` field which is used to store the serialized block markup](https://github.com/WordPress/gutenberg/blob/7fcd57c9a62 ...
```
It's cool that we want to document how this works today, but it is very likely to not continue down this path because it does not allow us to save attributes of the wrapping navigation block itself.

getdave replied:
Yup. Fair enough. I'd like to document correct workings for now as that encourages future contributors to write and update docs in order to correct things 😄

getdave replied:
Also writing this helps to ensure _I_ fully understand things. If you can't explain it you don't understand it.

getdave committed:
Add technical reference to where Nav block rendering happens

getdave commented:
I haven't forgotten about this...

getdave committed:
Disambiguate "links"

Addresses https://github.com/WordPress/gutenberg/pull/34682#discussion_r705985150

getdave committed:
Use "Classic" terminology instead of "Default"

Mamaduka commented:
The update looks great, Dave.

The "Hooks" section is updated, but I can correct it in follow-up PR today/tomorrow.

The new central hooks are `useNavigationEditor` and `useNavigationBlockEditor`. The Layout component itself had many changes and probably won't fit in the small snippet we currently have in docs.

getdave committed:
Ditch technical implementation details

As requested by https://github.com/WordPress/gutenberg/pull/34682#discussion_r712986069

getdave commented:
Seeking final ✅ on the basis that the scope of this PR is to 

* correct inconsistencies
* correct errors

🙏 😄

Mamaduka reviewed (approved):
I think this PR is in good shape, and all feedback got addressed 🌟 

I'll do a follow-up for hooks tomorrow.
